services/src/main/scala/cromwell/pipeline/service/ProjectService.scala

```diff
@@ -4,26 +4,32 @@ import java.util.UUID
 
 import cromwell.pipeline.datastorage.dao.repository.ProjectRepository
 import cromwell.pipeline.datastorage.dto._
-import cromwell.pipeline.service.Exceptions.{ ProjectAccessDeniedException, ProjectNotFoundException }
 import cromwell.pipeline.model.wrapper.UserId
+import cromwell.pipeline.service.Exceptions.{
+  ProjectAccessDeniedException,
+  ProjectIllegalStateException,
+  ProjectNotFoundException
+}
 
 import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success, Try }
 
 class ProjectService(projectRepository: ProjectRepository, projectVersioning: ProjectVersioning[VersioningException])(
   implicit executionContext: ExecutionContext
 ) {
 
-  def getProjectById(projectId: ProjectId): Future[Option[Project]] =
-    projectRepository.getProjectById(projectId)
+  private[service] def getProjectById(projectId: ProjectId, userId: UserId): Future[Project] =
+    projectRepository.getProjectById(projectId).map(getForUserOrFail(_, userId)).flatMap(Future.fromTry)
+
+  def getProjectByName(namePattern: String, userId: UserId): Future[Project] =
+    projectRepository.getProjectByName(namePattern).map(getForUserOrFail(_, userId)).flatMap(Future.fromTry)
 
-  def getProjectByName(namePattern: String, userId: UserId): Future[Option[Project]] = {
-    val result = projectRepository.getProjectByName(namePattern)
-    result.flatMap {
-      case Some(project) if project.ownerId == userId => result
-      case Some(_)                                    => Future.failed(new ProjectAccessDeniedException)
-      case None                                       => Future.failed(new ProjectNotFoundException)
+  private def getForUserOrFail(project: Option[Project], userId: UserId): Try[Project] =
```

Comment:
why here we switched to Try? If you continue with Future[Project] it will be less transformations (Future.fromTry), so like

```
 project match {
      case Some(project) if project.ownerId == userId => Future.successful(project)
      ...
```

Reply:
It's just my personal preference. If we have `Try`, we can easily reuse it and integrate into a code that uses `Future` or `Option`. If we have `Future`, it's not that easy to integrate it with `Try` or `Option`.
But since it is a private method and these transformations are ugly, I think Future[Project] should be fine too.